### PR TITLE
Implement Scala 2.13-3.0 sandwich

### LINKED ISSUE #5369: Provide a simple way for a local project to depend on another local project with a different scalaBinaryVersion (important for dotty/scala2 interop)
Dotty can read Scala 2 signatures (and soon the reverse will be true too), so a dotty project can depend on a scala 2 project, sbt-dotty makes this as convenient as possible by providing a `withDottyCompat` method to change the cross version:
```scala
libraryDependencies += ("a" %% "b" % "c").withDottyCompat(scalaVersion.value)
```
However, when using `dependsOn` to declare a local dependency between two projects, there is no way to specify the cross version, this prevents having a local dotty project that depends on a local scala2 project and vice-versa, even though this would be extremely useful for progressively porting a project to dotty.

As a workaround, one can fiddle with `projectDependencies`, but it's not very nice:
```scala
lazy val scala2Lib = project.in(file("scala2Lib"))
  .settings(
    scalaVersion := "2.13.1"
  )

lazy val dottyApp = project.in(file("dottyApp"))
  .dependsOn(scala2Lib)
  .settings(
    // Set withDottyCompat on the scala2Lib dependency if we had a mix of dotty
    // and scala2 dependencies, we would have to be more careful here.
    projectDependencies := {
      projectDependencies.value.map(_.withDottyCompat(scalaVersion.value))
    },
    scalaVersion := "0.21.0-RC1"
  )
```

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes https://github.com/sbt/sbt/issues/5369
Ref https://contributors.scala-lang.org/t/roadmap-for-the-tasty-reader-for-scala-2/4231

This implements support for inter-project dependencies between Scala 2.13 and Dotty, and vice versa. Scala 2.13 depending on Dotty would require 2.13.4 and above.

```scala
ThisBuild / scalaVersion := "0.23.0"

ThisBuild / resolvers += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"
// TODO use 2.13.4 when it's out
lazy val scala213 = "2.13.4-bin-aeee8f0"

lazy val root = (project in file("."))
  .aggregate(fooApp, fooCore, barApp, barCore)

lazy val fooApp = (project in file("foo-app"))
  .dependsOn(fooCore)

lazy val fooCore = (project in file("foo-core"))
  .settings(
    scalaVersion := scala213
  )

lazy val barApp = (project in file("bar-app"))
  .dependsOn(barCore)
  .settings(
    scalaVersion := scala213
  )

lazy val barCore = (project in file("bar-core"))
```

### CODE REVIEW

**Comment by eed3si9n** [comment]
`plugins/dotty-sandwich` works on my machine :) but it's erring out on Travis CI:

```
[info] Running plugins/dotty-sandwich
[info] [info] welcome to sbt 1.4.0-SNAPSHOT (AdoptOpenJDK Java 11.0.1)
[info] [info] loading settings for project global-plugins from metals.sbt ...
[info] [info] loading global plugins from /tmp/sbt_d7daa57/global/plugins
[info] [info] loading settings for project sbt_d7daa57-build from plugins.sbt ...
[info] [info] loading project definition from /tmp/sbt_d7daa57/project
[info] [info] loading settings for project root from build.sbt ...
[info] [info] set current project to root (in build file:/tmp/sbt_d7daa57/)
[info] [info] compiling 1 Scala source to /tmp/sbt_d7daa57/foo-core/target/scala-2.13/classes ...
[info] [info] compiling 1 Scala source to /tmp/sbt_d7daa57/bar-core/target/scala-0.23/classes ...
[info] [info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.4-bin-aeee8f0. Compiling...
[info] [info] done compiling
[info] [info] compiling 1 Scala source to /tmp/sbt_d7daa57/bar-app/target/scala-2.13/classes ...
[info] [info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.4-bin-aeee8f0. Compiling...
[info] [info]   Compilation completed in 18.539s.
[info] [info]   Compilation completed in 24.223s.
[info] [info] Attempting to fetch org.scala-sbt:compiler-bridge_2.13:1.4.0-M10.
[info] [error] ## Exception when compiling 1 sources to /tmp/sbt_d7daa57/foo-core/target/scala-2.13/classes
[info] [error] sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$1$1: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: Cannot redefine component. ID: sbt.internal.scriptedtest.ScriptedLauncher$1@73f23e8e, files: /tmp/sbt_159a0475/org.scala-sbt-compiler-bridge_2.13-1.4.0-M10-bin_2.13.4-bin-aeee8f0__55.0.jar
[info] [error] 
[info] [error]            
[info] [info] done compiling
[info] [error] sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$1$1: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: Cannot redefine component. ID: sbt.internal.scriptedtest.ScriptedLauncher$1@73f23e8e, files: /tmp/sbt_159a0475/org.scala-sbt-compiler-bridge_2.13-1.4.0-M10-bin_2.13.4-bin-aeee8f0__55.0.jar
[info] [error] Caused by: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: Cannot redefine component. ID: sbt.internal.scriptedtest.ScriptedLauncher$1@73f23e8e, files: /tmp/sbt_159a0475/org.scala-sbt-compiler-bridge_2.13-1.4.0-M10-bin_2.13.4-bin-aeee8f0__55.0.jar
[info] [error] (fooCore / Compile / compileIncremental) sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$1$1: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: sbt.internal.scriptedtest.ScriptedLauncher$3$1$1$1$foo: Cannot redefine component. ID: sbt.internal.scriptedtest.ScriptedLauncher$1@73f23e8e, files: /tmp/sbt_159a0475/org.scala-sbt-compiler-bridge_2.13-1.4.0-M10-bin_2.13.4-bin-aeee8f0__55.0.jar
[error] x plugins/dotty-sandwich 
[error]  Cause of test exception: {line 1}  Command failed: compile failed
```

Looks like compiler bridge is being compiled twice?

File: main/src/main/scala/sbt/Defaults.scala [hunk 1]
```diff
@@ -3450,14 +3450,38 @@ object Classpaths {
...
+                depProjId.withConfigurations(dep.configuration).withExplicitArtifacts(Vector.empty)
+              )
+            else if (isScala2Scala3Sandwich(sbv, depSBV) && depCross == CrossVersion.binary)
+              Some(
+                depProjId
+                  .withCrossVersion(CrossVersion.constant(depSBV))
+                  .withConfigurations(dep.configuration)
+                  .withExplicitArtifacts(Vector.empty)
+              )
+            else sys.error(s"scalaBinaryVersion mismatch: expected $sbv but found ${depSBV}")
```
Comment (smarter) [inline main/src/main/scala/sbt/Defaults.scala:3483]: Maybe the error could mention the names of the mismatched projects?

File: sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+lazy val root = (project in file("."))
+  .aggregate(fooApp, fooCore, barApp, barCore)
+
+lazy val fooApp = (project in file("foo-app"))
+  .dependsOn(fooCore)
+
+lazy val fooCore = (project in file("foo-core"))
+  .settings(
+    scalaVersion := scala213
```
Comment (smarter) [inline sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt:15]: For completeness sake, I think it'd be nice to have one or more tests involving crossScalaVersion and running ++compile, e.g. what happens if I have a scala 3 app that depends on a lib that cross-compiles between 2 and 3, and vice-versa.

File: main/src/main/scala/sbt/Defaults.scala (hunk 1 above)
Comment (eed3si9n) [reply main/src/main/scala/sbt/Defaults.scala:3483]: The project name should show up as `projectDependencies` task failure, I think.

File: sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt (hunk 2 above)
Comment (eed3si9n) [reply sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt:15]: That's probably not gonna work.

Comment (er1c) [reply sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt:15]: Am I misreading previous @smarter comments? A 2.13.4+ (not released yet) should be able to depend on a dotty library, so why wouldn't this visa-versa test work?

Comment (eed3si9n) [reply sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt:15]: `+` command is ignorant about the sandwich, so if there is a list of scalaVersions it wouldn't know how to align them. I'd much rather handle that scenario using sbt-projectmatrix - https://github.com/sbt/sbt-projectmatrix/issues/28

Comment (er1c) [reply sbt/src/sbt-test/plugins/dotty-sandwich/build.sbt:15]: thanks for explaining!  that makes sense!

**Pull request merged**